### PR TITLE
Simplify Compiler Error Emission and Fix `slice2swift` Metadata Check

### PR DESCRIPTION
**Pull request by InsertCreativityHere**
This PR does 2 things:

1) There are 2 paths to report an error in the compiler: through `Unit->error` and through `DefinitionContext->error`.
That second path is used in only 2 places... So I deleted it. Now there's a single path for reporting compiler errors.
This also lets us delete `CompilerException` (which of course inherited from `Ice::LocalException`).

2) While fixing that, I noticed a bug in `slice2swift`'s metadata validation introduced in #1627.
TL;DR we had a `validate` function, with a bunch of parameters. The 2nd to last was `bool local`.
This bool was removed from the call-sites, but not from the function itself, so the 'last' parameter was acidentally being passed in for that `local` bool.

### CODE REVIEW

File: cpp/src/slice2swift/SwiftUtil.h [hunk 1]
```diff
@@ -168,13 +168,7 @@ namespace Slice
...
 
         private:
-            StringList validate(
-                const SyntaxTreeBasePtr&,
-                const StringList&,
-                const std::string&,
-                int,
-                bool local = false,
-                bool operationParameter = false);
+            StringList validate(const SyntaxTreeBasePtr&, const StringList&, const std::string&, int);
```
Comment (bernardnormier) [inline cpp/src/slice2swift/SwiftUtil.h:171]: Please add parameter names whenever you update a function declaration.

File: cpp/src/slice2swift/SwiftUtil.cpp [hunk 2]
```diff
@@ -1803,7 +1801,7 @@ SwiftGenerator::MetadataVisitor::visitModuleStart(const ModulePtr& p)
             ostringstream os;
             os << "invalid module mapping:\n Slice module `" << p->scoped() << "' should be map to Swift module `"
                << current->second << "'" << endl;
-            dc->error(p->file(), p->line(), os.str());
+            unit->error(p->file(), p->line(), os.str());
```
Comment (bernardnormier) [inline cpp/src/slice2swift/SwiftUtil.cpp:1804]: The error message above is not grammatically correct.

Also, for my understanding, what's the difference between p->file()/p->line() and the overload without them?

File: cpp/src/slice2swift/SwiftUtil.cpp [hunk 3]
```diff
@@ -2538,21 +2536,19 @@ SwiftGenerator::MetadataVisitor::visitOperation(const OperationPtr& p)
         }
     }
     p->setMetadata(validate(p, metadata, p->file(), p->line()));
-    ParamDeclList params = p->parameters();
-    for (ParamDeclList::iterator q = params.begin(); q != params.end(); ++q)
+    for (auto& param : p->parameters())
```
Comment (bernardnormier) [inline cpp/src/slice2swift/SwiftUtil.cpp:2539]: const auto&

File: cpp/src/slice2swift/SwiftUtil.cpp [hunk 4]
```diff
@@ -2538,21 +2536,19 @@ SwiftGenerator::MetadataVisitor::visitOperation(const OperationPtr& p)
...
     }
 }
 
 bool
 SwiftGenerator::MetadataVisitor::visitExceptionStart(const ExceptionPtr& p)
 {
     p->setMetadata(validate(p, p->getMetadata(), p->file(), p->line()));
-    DataMemberList members = p->dataMembers();
-    for (DataMemberList::iterator q = members.begin(); q != members.end(); ++q)
+    for (auto& member : p->dataMembers())
```
Comment (bernardnormier) [inline cpp/src/slice2swift/SwiftUtil.cpp:2549]: Ditto

File: cpp/src/slice2swift/SwiftUtil.cpp [hunk 5]
```diff
@@ -2591,7 +2586,11 @@ SwiftGenerator::MetadataVisitor::visitDictionary(const DictionaryPtr& p)
             continue;
         }
 
-        dc->error(p->file(), p->line(), "invalid metadata `" + s + "' for dictionary key type");
+        dc->warning(
+            InvalidMetadata,
+            p->file(),
+            p->line(),
+            "ignoring invalid metadata `" + s + "' for dictionary key type");
```
Comment (bernardnormier) [inline cpp/src/slice2swift/SwiftUtil.cpp:2593]: ` => '

**Review by bernardnormier — APPROVED** [review]

File: cpp/src/Slice/FileTracker.h [hunk 6]
```diff
@@ -3,6 +3,7 @@
 #ifndef FILE_TRACKER_H
 #define FILE_TRACKER_H
 
+#include "Ice/LocalException.h"
```
Comment (externl) [inline cpp/src/Slice/FileTracker.h:6]: Is this needed?

**Review by externl — APPROVED** [review]

File: cpp/src/Slice/FileTracker.h (hunk 6 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/FileTracker.h:6]: Yes, we define a `FileException` which inherits from `LocalException`.
We used to get this include transitively because `Parser.h` included it, but now it doesn't.

File: cpp/src/slice2swift/SwiftUtil.cpp (hunk 5 above)
Comment (InsertCreativityHere) [reply cpp/src/slice2swift/SwiftUtil.cpp]: Fixed.

File: cpp/src/slice2swift/SwiftUtil.cpp (hunk 4 above)
Comment (InsertCreativityHere) [reply cpp/src/slice2swift/SwiftUtil.cpp]: Ditto

File: cpp/src/slice2swift/SwiftUtil.cpp (hunk 3 above)
Comment (InsertCreativityHere) [reply cpp/src/slice2swift/SwiftUtil.cpp]: Changed.

I was just keeping the original semantics; we used `iterator` and not `const_iterator`,
probably because within the loop, we were calling `setMetadata` which mutates the `param` (but through a ptr).

File: cpp/src/slice2swift/SwiftUtil.cpp (hunk 2 above)
Comment (InsertCreativityHere) [reply cpp/src/slice2swift/SwiftUtil.cpp:1804]: If you don't provide a `file` and `line` parameter, the parser/`Unit` will use `currentFile()` and `currentLine()` for them. This is useful _while_ parsing is occurring, since these are almost always what we want to pass.

Once we hit the visitors though, parsing has already finished, so `currentFile` and `currentLine` will be the _last_ file and line the parser saw, which isn't what we want.

----

Also I fixed the error message.
